File: algo/18. 4Sum.py
class Solution:
    def fourSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        
        # Time complexity: O(n^3)

        if len(nums) < 4:
            return []

        nums.sort()

        res = dict()

        for i in range(len(nums)-3):
            for j in range(i+1, len(nums)-2):
                left = j+1
                right = len(nums)-1
                num_i = nums[i]
                num_j = nums[j]
                #print("i =",i, "j =", j)
                while left != right:
                    num_left = nums[left]
                    num_right = nums[right]
                    currSum = num_i + num_j + num_left + num_right
                    if currSum == target and (num_i, num_j, num_left, num_right) not in res.keys():
                        res[(num_i, num_j, num_left, num_right)] = 1
                        left += 1
                    elif currSum == target:
                        left += 1
                    elif currSum < target:
                        left += 1
                    elif currSum > target:
                        right -= 1
                    #print("left = ",left," right =",right)
        return list(res.keys())



print(Solution().fourSum([1,0,-1,0,-2,2], 0))

File: round2/18. 4Sum.cs
public class Solution {
    public IList<IList<int>> FourSum(int[] nums, int target) {
        var res = new List<IList<int>>();
        if (nums.Length < 4) return res;
        Array.Sort(nums);
        for (int i=0; i<nums.Length-3; i++) {
            if (i>0 && nums[i] == nums[i-1]) continue;
            for (int j=i+1; j<nums.Length-2; j++) {
                if (j>i+1 && nums[j]==nums[j-1]) continue;
                int left = j+1, right = nums.Length-1;
                while (left < right) {
                    int currSum = nums[i]+nums[j]+nums[left]+nums[right];
                    if (currSum == target) {
                        res.Add(new List<int>(){nums[i],nums[j],nums[left],nums[right]});
                        left ++;
                        right --;
                        while (left < right && nums[left] == nums[left-1]) {
                            left ++;
                        }
                        while (left < right && nums[right] == nums[right+1]) {
                            right --;
                        }
                    } else if (currSum < target) {
                        left ++;
                    } else {
                        right --;
                    }
                }
            }
        }
        return res;
    }
}